src/main/java/RecommendationsServlet.java

```diff
@@ -0,0 +1,184 @@
+                .forEachOrdered(x -> reverseSortedMap.put(x.getKey(), x.getValue()));
+    return reverseSortedMap;
+  }
+
+  // Return suggested recommendations based on the history. Create hashmaps for each input for all of the queries,
+  // then create four combinations based on the most frequented fields.
+  private ArrayList<VisualizationData> getRecommendations(ArrayList<VisualizationData> userHistory, String userId) {
+    // Populate the frequency maps for each field
+    getFrequencyMaps(userHistory);
+    // Sort the frequency maps in descending order by value such that the most common searches for each field 
+    // are first eg: {worked in: 7, lived in: 10, moved to: 2} => { lived in: 10, worked in: 7, moved to: 2}
+    LinkedHashMap<String, Integer> reverseSortedPersonType = sortHashMapDescending(personTypeMap);
+    LinkedHashMap<String, Integer> reverseSortedAction = sortHashMapDescending(actionMap);
+    LinkedHashMap<String, Integer> reverseSortedLocation = sortHashMapDescending(locationMap);
+    LinkedHashMap<String, Integer> reverseSortedYear = sortHashMapDescending(yearsMap);
+    // Get the keyset from each frequency map with the guaranteed order from sorting
+    Set<String> personTypeKeys = reverseSortedPersonType.keySet();
+    Set<String> actionKeys = reverseSortedAction.keySet();
+    Set<String> locationKeys = reverseSortedLocation.keySet();
+    Set<String> yearKeys = reverseSortedYear.keySet();
+    ArrayList<VisualizationData> recommendationList = new ArrayList<VisualizationData>();
+    // Iterate through all fields and find combinations that the user hasn't made yet
+    // and return the top 3.
+    for (String personType : personTypeKeys) {
+      for(String action : actionKeys) {
+        for (String location : locationKeys) {
+          for (String year : yearKeys) {
+            VisualizationData recommendation = new VisualizationData(userId, personType, action, location, year);
+             if (isRecommendationValid(recommendation, userHistory)) {
+               recommendationList.add(recommendation);
+               if (recommendationList.size() > 3) {
+                 return recommendationList;
+               }
+             }
+          }
+        }
+      }
+    }
+    return recommendationList;
+  }
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String userId = request.getParameter("user-id");
+    History userHistoryObj = new History(userId);
+    ArrayList<VisualizationData> userHistory = userHistoryObj.getHistoryList();
+    if (userHistory.isEmpty()) {
+      String json = new Gson().toJson(getDefaultRecommendations(userId));
+      response.getWriter().write(json);
+    } else {
+      ArrayList<VisualizationData> recommendations = getRecommendations(userHistory, userId);
+      // If the recommendation algorithm is unable to generate more than 4 possibilities
```

Comment:
Adding to previous comment, I seem to always get four from the getRecommendations function and 1 random

Reply:
Yes, for the most part you should always get four from getRecommendations and 1 random except if you have many queries with only one field changed. Eg: "How many people lived in each California County in 2018" and "How many people lived in each Florida County in 2018". With these two elements in the history, there are no new combinations we can make, so more random ones are returned. Ideally the algorithm would return more visualizations about the existing states, which I'll implement if I have time.

Reply:
Oh I see, thanks!